feat: reindex-algolia constant memory consumption

pwnage101 opened the pull request:
Before this change, the `index_enterprise_catalog_in_algolia_task` task (a.k.a. reindex-algolia) had a maximum memory consumption propotional to the total number of content_keys and number of related objects such as enterprises, catalogs, catalog queries, etc.  After this change, the maximum memory consumption will no longer be sensitive to the number of content keys (but still sensitive to the number of related objects).  This should significantly reduce the memory footprint of the task.

Estimate: In prod recently, I have observed 95,117 algolia products being indexed in total.  The default algolia library batch size is 1,000 products, so this should result in a ~95x factor decrease in memory consumption.

This is achieved by leveraging the fact that Algolia's `replace_all_objects` library method accepts any iterable, including generators, and does not accumulate more than 1 batch in-memory before shipping it over the network.  Before this change, we were accumulating all results into a list before even calling `replace_all_objects`; after this change, we instead construct a lazy generator and pass that along.

ENT-7311

pwnage101 commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,11 +555,147 @@ def add_metadata_to_algolia_objects(
 
-def index_content_keys_in_algolia(content_keys, algolia_client):  # pylint: disable=too-many-statements
+# pylint: disable=too-many-statements
+def _get_algolia_products_for_batch(
```
The vast majority of this function is just copy+paste existing code.  Only changes are minor variable name changes.

pwnage101 commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -692,9 +736,9 @@ def index_content_keys_in_algolia(content_keys, algolia_client):  # pylint: disa
-    # extract out only the fields we care about and send to Algolia index
-    algolia_objects = create_algolia_objects(algolia_products_by_object_id.values(), ALGOLIA_FIELDS)
-    algolia_client.replace_all_objects(algolia_objects)
+    # Feed the un-evaluated flat iterable of algolia products into the 3rd party library function.  As of this writing,
+    # this library function will chunk the interable again using a default batch size of 100.
+    algolia_client.replace_all_objects(algolia_products_generator)
```
This is basically where the magic happens.  Instead of passing a huge 16GB+ python object into this function, we pass a 0GB generator.  `replace_all_objects` happens to already be written in a way that it will not load the entire list into memory.

adamstankiewicz commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,147 +555,188 @@ def add_metadata_to_algolia_objects(
-        )
-        # iterate over courses, programs and pathways and add their metadata to the list of objects to be indexed
-        for metadata in filtered_content_metadata:
+    # iterate through ContentMetadata records, retrieving the enterprise_catalog_uuids
+    # and enterprise_customer_uuids associated with each ContentMetadata record (either
+    # a course or a course run or program), storing them in a dictionary with the related
+    # content_key as a key for later retrieval. the content_key is determined by
+    # the content_key field if the metadata is a `COURSE` or `PROGRAM` or by the parent_content_key
+    # field if the metadata is a `COURSE_RUN`.
```
nit: unrelated to your changes, also retrieves the associated `enterprise_catalog_queries` as well now

adamstankiewicz commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,147 +555,188 @@ def add_metadata_to_algolia_objects(
+                customer_uuids_by_key[pathway.content_key].update(enterprise_customer_uuids)
+
+    # iterate through the courses and programs, retrieving the enterprise-related uuids from the
+    # dictionary created above. there is at least 2 duplicate course records per course,
```
nit: "at least **3** duplicate course records records per course", since it has the `catalog_queries_by_key` (i.e., catalog query uuids), too, now, though unrelated to your immediate changes.

adamstankiewicz commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,147 +555,188 @@ def add_metadata_to_algolia_objects(
+    # iterate over courses, programs and pathways and add their metadata to the list of objects to be indexed
+    for metadata in filtered_content_metadata:
+        content_key = metadata.content_key
+        # Check if we've indexed the course recently
+        # (programs/pathways are indexed every time regardless of last indexing)
```
[unrelated to your change, general musing] I wonder why this is? 🤔

pwnage101 replied:
The explanation is here:

https://github.com/openedx/enterprise-catalog/blob/f5c761158f0622cc4f96d1427268485189326bb7/enterprise_catalog/apps/api/tasks.py#L788-L798

but it feels a bit sus to me because it seems to be describing functionality that doesn't need to leverage a caching framework.  I'm not putting too much thought into it though since I'm treating it as out of scope for this PR.

iloveagent57 replied:
But on the other hand - we probably _do_ want to update the record in that case, because it likely has new catalog or query uuids associated with it.  So yeah, I think we probably don't want this `_was_recently_indexed()` call at all.  But don't worry about it for this PR.

adamstankiewicz commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,147 +555,188 @@ def add_metadata_to_algolia_objects(
+        'total_products_count': 0,
+    }
+    # Produce a generator of batches of algolia products to index.  Each batch has an unpredictable, variable length.
+    # Not immediately evaluated, so no memory is consumed yet.
```
🎉

adamstankiewicz commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,147 +555,188 @@ def add_metadata_to_algolia_objects(
+
+    # Feed the un-evaluated flat iterable of algolia products into the 3rd party library function.  As of this writing,
+    # this library function will chunk the interable again using a default batch size of 1000.
+    algolia_client.replace_all_objects(algolia_products_generator)
```
nit: I wonder if its worth leaving a link to the Algolia docs for this `replace_all_objects` method as a comment would be helpful for future us?

adamstankiewicz commented on enterprise_catalog/apps/api/tasks.py:
```diff
@@ -555,147 +555,188 @@ def add_metadata_to_algolia_objects(
+    catalog_queries = CatalogQuery.objects.prefetch_related(
+        'enterprise_catalogs',
+    )
+    associate_programs_query = ContentMetadata.objects.filter(content_type__in=[PROGRAM, LEARNER_PATHWAY])
```
nit: unrelated to your immediate changed, but i believe this variable should be `associated_programs_query` (with the `d`)?

adamstankiewicz reviewed (approved):
these changes look good to me! all my comments are nits, mostly about prior code / namings.

pwnage101 commented:
Refactored the function signatures a bit to be more sane/readable, overhauled the comments and moved them closer to the code they were describing, added a test for the new generator logic.
